dask/dataframe/groupby.py

```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+# 2. collect all partitions into a single partition
+# 3. group-by-aggregate the result into intermediate values
+# 4. transform all intermediate values into the result
+#
+# In Step 1 and 3 the dataframe is grouped on the same columns.
+#
+###############################################################
+
+def _aggregate_meta(df, index, levels, chunk_funcs, agg_funcs, finalizers):
+    stage1 = _groupby_apply_funcs(df, funcs=chunk_funcs, by=index)
+    stage2 = _groupby_apply_funcs(stage1, funcs=agg_funcs, level=levels)
+    return _agg_finalize(stage2, finalizers)
+
+
+def _normalize_spec(spec, non_group_columns):
+    """
+    Return a list of ``(result_column, func, input_column)`` tuples.
+    """
+    if not isinstance(spec, dict):
+        spec = collections.OrderedDict(zip(non_group_columns, it.repeat(spec)))
+
+    res = []
+
+    if isinstance(spec, dict):
+        for input_column, subspec in spec.items():
+            if isinstance(subspec, dict):
+                res.extend(((input_column, result_column), func, input_column)
+                           for result_column, func in subspec.items())
+
+            else:
+                if not isinstance(subspec, list):
+                    subspec = [subspec]
+
+                res.extend(((input_column, funcname(func)), func, input_column)
+                           for func in subspec)
+
+    else:
+        raise ValueError("unsupported agg spec of type {}".format(type(spec)))
+
+    compounds = (list, tuple, dict)
+    use_flat_columns = not any(isinstance(subspec, compounds)
+                               for subspec in spec.values())
+
+    if use_flat_columns:
+        res = [(input_col, func, input_col) for (_, func, input_col) in res]
+
+    return res
+
+
+def _build_agg_args(spec):
+    """
+    Create the transformation functions for a normalized aggregate spec.
+
+    Returns
+    -------
+    chunk_funcs: a list of functions that are applied on grouped chunks of the
+        initial dataframe. Each function takes a grouped dataframe and should
+        return a dataframe of intermediate values with the group keys as an
+        index.
+
+    agg_funcs: a list of functions that are applied on the grouped dataframe
+        after all chunks have been concatinated. Each function takes a grouped
+        dataframe and should return a dataframe of intermediate values with the
+        group keys as an index.
+
+    finalizers: a list of result-column transform pairs, that are applied after
+        the ``agg_funcs``. They are used to create final results from
+        intermediate representations.
+    """
+    # generator for consecutive IDs (for intermediate results)
+    ids = it.count()
+    next_id = lambda prefix: '{}-{}'.format(prefix, next(ids))
+
+    simple_impl = {
+        'sum': (M.sum, M.sum),
+        'min': (M.min, M.min),
+        'max': (M.max, M.max),
+        'count': (M.count, M.sum),
+        'size': (M.size, M.sum),
+    }
```

Comment:
For all aggregation supported `combine == aggregate`. While, I would be happy to add the corresponding tuple elements, dictionary entries in the `_build_ ...` functions, I feel it would needlessly clutter the code. However, I think adding the corresponding keyword arguments for `apply_concat_apply` would make the assumption of `combine == aggregate` explicit.